File: server/app/utils/pool_leak_detector.py
"""Bug-94: 数据库连接池泄漏检测.

设计:
  - 跟踪每次 checkout / checkin
  - 超 N 分钟未归还视为泄漏
  - 提供 force_release 接口 (后端 worker 自动回收)
  - 配合 SQLAlchemy Pool event hooks
  - 记录泄漏栈 + 调用者

使用:
    from app.utils.pool_leak_detector import pool_leak_detector, install_sa_pool_hook

    install_sa_pool_hook(engine1)
    pool_leak_detector.set_timeout(300)  # 5min
"""

import logging
import threading
import time
import traceback
from collections import deque
from dataclasses import dataclass
from typing import Any

logger = logging.getLogger(__name__)

DEFAULT_LEAK_TIMEOUT_SEC = 300.0
DEFAULT_MAX_RECORDS = 200


@dataclass
class CheckOutRecord:
    conn_id: int
    engine: str
    checked_out_at: float
    stack: str
    context: str
    checked_in_at: float = 0.0

    def age_sec(self) -> float:
        if self.checked_in_at:
            return self.checked_in_at - self.checked_out_at
        return time.time() - self.checked_out_at

    def to_dict(self) -> dict:
        return {
            "conn_id": self.conn_id,
            "engine": self.engine,
            "checked_out_at": round(self.checked_out_at, 3),
            "checked_in_at": round(self.checked_in_at, 3) if self.checked_in_at else None,
            "age_sec": round(self.age_sec(), 1),
            "context": self.context[:100],
            "stack": self.stack[:300],
        }


class PoolLeakDetector:
    """DB 连接池泄漏检测器."""

    def __init__(self, leak_timeout_sec: float = DEFAULT_LEAK_TIMEOUT_SEC, max_records: int = DEFAULT_MAX_RECORDS):
        self._lock = threading.Lock()
        self._timeout = leak_timeout_sec
        self._max_records = max_records
        self._outstanding: dict[int, CheckOutRecord] = {}  # conn_id -> record
        self._history: deque[CheckOutRecord] = deque(maxlen=max_records)
        self._leak_warnings: deque[CheckOutRecord] = deque(maxlen=max_records)
        self._next_id = 1
        self._total_checkout = 0
        self._total_checkin = 0
        self._total_leaked = 0
        self._total_force_released = 0

    def set_timeout(self, sec: float) -> None:
        with self._lock:
            self._timeout = max(1.0, float(sec))

    def _new_id(self) -> int:
        with self._lock:
            i = self._next_id
            self._next_id += 1
            return i

    def checkout(self, engine: str, context: str = "") -> int:
        """记录一次 checkout. 返回 conn_id (>=1)."""
        conn_id = self._new_id()
        rec = CheckOutRecord(
            conn_id=conn_id,
            engine=engine,
            checked_out_at=time.time(),
            stack="".join(traceback.format_stack(limit=8)),
            context=context,
        )
        with self._lock:
            self._outstanding[conn_id] = rec
            self._total_checkout += 1
        return conn_id

    def checkin(self, conn_id: int) -> None:
        """归还连接. 若 id 不存在则忽略."""
        with self._lock:
            rec = self._outstanding.pop(conn_id, None)
            if rec is not None:
                rec.checked_in_at = time.time()
                self._history.append(rec)
                self._total_checkin += 1
            else:
                # 可能是 force_released 过了, 静默忽略
                pass

    def scan_leaks(self) -> list[dict[str, Any]]:
        """扫描超时未归还的连接, 记录为 leak. 返回新发现的 leak 列表."""
        now = time.time()
        new_leaks: list[CheckOutRecord] = []
        with self._lock:
            th = self._timeout
            for _cid, rec in list(self._outstanding.items()):
                if now - rec.checked_out_at > th:
                    new_leaks.append(rec)
                    self._leak_warnings.append(rec)
                    self._total_leaked += 1
        for rec in new_leaks:
            logger.warning(
                f"pool_leak: conn_id={rec.conn_id} engine={rec.engine} "
                f"age={now - rec.checked_out_at:.0f}s > {th}s context={rec.context[:80]}"
            )
        return [r.to_dict() for r in new_leaks]

    def force_release(self, conn_id: int) -> bool:
        """强制回收超时连接. True 表示已回收."""
        with self._lock:
            rec = self._outstanding.pop(conn_id, None)
            if rec is None:
                return False
            rec.checked_in_at = time.time()
            self._total_force_released += 1
            self._history.append(rec)
            return True

    def force_release_all_leaked(self) -> int:
        """回收所有超时连接. 返回回收数."""
        ids = [cid for cid, r in self._outstanding.items() if time.time() - r.checked_out_at > self._timeout]
        n = 0
        for cid in ids:
            if self.force_release(cid):
                n += 1
        return n

    def get_outstanding(self) -> list[dict[str, Any]]:
        with self._lock:
            return [r.to_dict() for r in self._outstanding.values()]

    def get_leaks(self) -> list[dict[str, Any]]:
        with self._lock:
            return [r.to_dict() for r in list(self._leak_warnings)[-50:][::-1]]

    def stats(self) -> dict:
        with self._lock:
            return {
                "outstanding": len(self._outstanding),
                "leak_timeout_sec": self._timeout,
                "total_checkout": self._total_checkout,
                "total_checkin": self._total_checkin,
                "total_leaked": self._total_leaked,
                "total_force_released": self._total_force_released,
                "leak_rate": round(self._total_leaked / self._total_checkout, 4) if self._total_checkout else 0.0,
                "kept_history": len(self._history),
            }

    def clear(self) -> None:
        with self._lock:
            self._outstanding.clear()
            self._history.clear()
            self._leak_warnings.clear()
            self._total_checkout = 0
            self._total_checkin = 0
            self._total_leaked = 0
            self._total_force_released = 0


# 全局单例
pool_leak_detector = PoolLeakDetector()


def install_sa_pool_hook(engine, engine_name: str | None = None) -> None:
    """给 SQLAlchemy engine 装 hook."""
    try:
        from sqlalchemy import event
    except Exception as e:
        logger.debug(f"sa pool hook skip: {e!r}")
        return
    name = engine_name or str(getattr(engine, "url", "unknown"))

    @event.listens_for(engine, "checkout")
    def _checkout(dbapi_conn, conn_record, conn_proxy):
        try:
            context = f"pool_size={engine.pool.size()}"
            pool_leak_detector.checkout(name, context=context)
        except Exception as e:
            logger.debug(f"sa pool checkout hook: {e!r}")

    @event.listens_for(engine, "checkin")
    def _checkin(dbapi_conn, conn_record):
        # checkin 没有直接 conn_id, 这里用 conn_record 引用做匹配
        # 简化: 找 outstanding 中第一个匹配的 engine
        try:
            with pool_leak_detector._lock:
                target = None
                for cid, rec in pool_leak_detector._outstanding.items():
                    if rec.engine == name and rec.checked_in_at == 0.0:
                        target = cid
                        break
            if target is not None:
                pool_leak_detector.checkin(target)
        except Exception as e:
            logger.debug(f"sa pool checkin hook: {e!r}")

File: apps/api/src/utils/pool-leak-detector.ts
/**
 * 数据库连接池泄漏检测器。
 *
 * 迁移自旧架构 app/utils/pool_leak_detector.py 和 bug152_pool_monitor.py。
 *
 * 设计：
 * - 跟踪每次 checkout / checkin（WeakMap 记录 stack trace）
 * - 超 N 秒未归还视为泄漏
 * - 提供 forceRelease 接口（后端 worker 自动回收）
 * - 超时未归还告警
 *
 * 使用：
 *   const connId = poolLeakDetector.checkout('writer', 'SELECT users');
 *   try {
 *     // ... 使用连接 ...
 *   } finally {
 *     poolLeakDetector.checkin(connId);
 *   }
 *
 *   // 定时扫描泄漏
 *   const leaks = poolLeakDetector.scanLeaks();
 */

// =============================================================================
// 类型定义
// =============================================================================

/** 借用状态。 */
export type BorrowState = 'OK' | 'TIMEOUT' | 'POOL_EXHAUSTED'

/** 连接借用记录。 */
export interface CheckOutRecord {
  /** 连接 ID（自增） */
  connId: number
  /** 连接池名称 */
  pool: string
  /** 借出时间（Unix 毫秒） */
  checkedOutAt: number
  /** 借出时的调用栈 */
  stack: string
  /** 上下文描述（如 SQL 语句摘要） */
  context: string
  /** 归还时间（Unix 毫秒），0 表示尚未归还 */
  checkedInAt: number
}

/** 泄漏检测统计。 */
export interface PoolLeakStats {
  /** 当前未归还的连接数 */
  outstanding: number
  /** 泄漏超时阈值（秒） */
  leakTimeoutSec: number
  /** 累计借出次数 */
  totalCheckout: number
  /** 累计归还次数 */
  totalCheckin: number
  /** 累计泄漏次数 */
  totalLeaked: number
  /** 累计强制回收次数 */
  totalForceReleased: number
  /** 泄漏率（0-1） */
  leakRate: number
  /** 历史保留记录数 */
  keptHistory: number
}

/** 借用操作记录（用于 stats 分析）。 */
export interface BorrowRecord {
  state: BorrowState
  /** 等待耗时（毫秒） */
  waitMs: number
  /** 使用耗时（毫秒） */
  usedMs: number
  /** 时间戳（Unix 毫秒） */
  ts: number
}

// =============================================================================
// 常量
// =============================================================================

const DEFAULT_LEAK_TIMEOUT_SEC = 300 // 5 分钟
const DEFAULT_MAX_RECORDS = 200

// =============================================================================
// 连接池泄漏检测器
// =============================================================================

/**
 * 数据库连接池泄漏检测器。
 *
 * 使用 Map（connId → CheckOutRecord）跟踪借出连接，
 * 借出时记录调用栈，归还时移除记录。
 * 超时未归还的连接在 scanLeaks() 时被标记为泄漏。
 */
export class PoolLeakDetector {
  private readonly outstanding = new Map<number, CheckOutRecord>()
  private readonly history: CheckOutRecord[] = []
  private readonly leakWarnings: CheckOutRecord[] = []
  private readonly maxRecords: number
  private leakTimeoutSec: number
  private nextId = 1
  private totalCheckout = 0
  private totalCheckin = 0
  private totalLeaked = 0
  private totalForceReleased = 0

  constructor(
    leakTimeoutSec: number = DEFAULT_LEAK_TIMEOUT_SEC,
    maxRecords: number = DEFAULT_MAX_RECORDS,
  ) {
    this.leakTimeoutSec = Math.max(1, leakTimeoutSec)
    this.maxRecords = Math.max(10, maxRecords)
  }

  /** 设置泄漏超时阈值（秒）。 */
  setTimeout(sec: number): void {
    this.leakTimeoutSec = Math.max(1, sec)
  }

  /**
   * 记录一次连接借出。
   *
   * @param pool 连接池名称（如 'writer' / 'reader'）
   * @param context 上下文描述（如 SQL 语句摘要）
   * @returns 连接 ID（>=1），用于后续 checkin
   */
  checkout(pool: string, context: string = ''): number {
    const connId = this.nextId++
    const stack = this.captureStack(8)

    const rec: CheckOutRecord = {
      connId,
      pool,
      checkedOutAt: Date.now(),
      stack,
      context,
      checkedInAt: 0,
    }

    this.outstanding.set(connId, rec)
    this.totalCheckout++

    return connId
  }

  /**
   * 归还连接。若 connId 不存在则静默忽略（可能已被 forceRelease）。
   */
  checkin(connId: number): void {
    const rec = this.outstanding.get(connId)
    if (rec === undefined) return

    rec.checkedInAt = Date.now()
    this.outstanding.delete(connId)
    this.addToHistory(rec)
    this.totalCheckin++
  }

  /**
   * 扫描超时未归还的连接，记录为泄漏。
   *
   * @returns 新发现的泄漏记录列表
   */
  scanLeaks(): CheckOutRecord[] {
    const now = Date.now()
    const thresholdMs = this.leakTimeoutSec * 1000
    const newLeaks: CheckOutRecord[] = []

    for (const [connId, rec] of this.outstanding) {
      if (now - rec.checkedOutAt > thresholdMs) {
        newLeaks.push(rec)
        this.leakWarnings.push(rec)
        this.totalLeaked++

        console.warn(
          `[pool-leak] conn_id=${rec.connId} pool=${rec.pool} ` +
            `age=${Math.round((now - rec.checkedOutAt) / 1000)}s > ${this.leakTimeoutSec}s ` +
            `context=${rec.context.slice(0, 80)}`,
        )

        // 从 outstanding 中移除（视为已泄漏，不再跟踪归还）
        this.outstanding.delete(connId)
        // 忽略 connId 未使用警告（已通过迭代器删除）
        void connId
      }
    }

    // 限制 leakWarnings 长度
    if (this.leakWarnings.length > this.maxRecords) {
      this.leakWarnings.splice(0, this.leakWarnings.length - this.maxRecords)
    }

    return newLeaks
  }

  /**
   * 强制回收指定连接。
   *
   * @returns true 表示已回收，false 表示连接不存在（可能已归还或已回收）
   */
  forceRelease(connId: number): boolean {
    const rec = this.outstanding.get(connId)
    if (rec === undefined) return false

    rec.checkedInAt = Date.now()
    this.outstanding.delete(connId)
    this.addToHistory(rec)
    this.totalForceReleased++
    return true
  }

  /**
   * 回收所有超时连接。
   *
   * @returns 回收数量
   */
  forceReleaseAllLeaked(): number {
    const now = Date.now()
    const thresholdMs = this.leakTimeoutSec * 1000
    const ids: number[] = []

    for (const [connId, rec] of this.outstanding) {
      if (now - rec.checkedOutAt > thresholdMs) {
        ids.push(connId)
      }
    }

    let count = 0
    for (const id of ids) {
      if (this.forceRelease(id)) count++
    }

    return count
  }

  /** 获取当前未归还的连接列表（用于诊断）。 */
  getOutstanding(): CheckOutRecord[] {
    return [...this.outstanding.values()].map((r) => ({ ...r }))
  }

  /** 获取最近的泄漏告警列表（最多 50 条，按时间倒序）。 */
  getLeaks(): CheckOutRecord[] {
    return this.leakWarnings
      .slice(-50)
      .reverse()
      .map((r) => ({ ...r }))
  }

  /** 获取统计信息。 */
  stats(): PoolLeakStats {
    return {
      outstanding: this.outstanding.size,
      leakTimeoutSec: this.leakTimeoutSec,
      totalCheckout: this.totalCheckout,
      totalCheckin: this.totalCheckin,
      totalLeaked: this.totalLeaked,
      totalForceReleased: this.totalForceReleased,
      leakRate:
        this.totalCheckout > 0
          ? Math.round((this.totalLeaked / this.totalCheckout) * 10000) / 10000
          : 0,
      keptHistory: this.history.length,
    }
  }

  /** 清空所有记录和计数器。 */
  clear(): void {
    this.outstanding.clear()
    this.history.length = 0
    this.leakWarnings.length = 0
    this.totalCheckout = 0
    this.totalCheckin = 0
    this.totalLeaked = 0
    this.totalForceReleased = 0
  }

  // ===========================================================================
  // 内部方法
  // ===========================================================================

  /** 捕获调用栈（限制深度）。 */
  private captureStack(limit: number): string {
    const err = new Error()
    const stack = err.stack ?? ''
    // 按 \n 分割，跳过前 3 行（Error / captureStack / checkout）
    const lines = stack.split('\n').slice(3, 3 + limit)
    return lines.join('\n')
  }

  /** 添加到历史记录（限制长度）。 */
  private addToHistory(rec: CheckOutRecord): void {
    this.history.push(rec)
    if (this.history.length > this.maxRecords) {
      this.history.splice(0, this.history.length - this.maxRecords)
    }
  }
}

/** 全局单例。 */
export const poolLeakDetector = new PoolLeakDetector()

// =============================================================================
// WeakMap 辅助：为任意连接对象附加泄漏跟踪
// =============================================================================

/**
 * 使用 WeakMap 为连接对象附加借用记录（不阻止 GC）。
 *
 * 适用于 postgres.js 等库的连接对象：借出时调用 track，
 * 归还时调用 untrack，超时后可通过 getTracked 查看借用信息。
 */
const trackedConnections = new WeakMap<
  object,
  { connId: number; pool: string; checkedOutAt: number; stack: string }
>()

/**
 * 跟踪一个连接对象的借出。
 *
 * @param conn 连接对象（任意对象引用）
 * @param pool 连接池名称
 * @param context 上下文描述
 * @returns 连接 ID
 */
export function trackConnection(conn: object, pool: string, context: string = ''): number {
  const connId = poolLeakDetector.checkout(pool, context)
  const stack = new Error().stack ?? ''
  trackedConnections.set(conn, {
    connId,
    pool,
    checkedOutAt: Date.now(),
    stack,
  })
  return connId
}

/**
 * 取消跟踪连接对象（归还时调用）。
 */
export function untrackConnection(conn: object): void {
  const info = trackedConnections.get(conn)
  if (info) {
    poolLeakDetector.checkin(info.connId)
    trackedConnections.delete(conn)
  }
}

/**
 * 获取被跟踪连接的借用信息（用于诊断）。
 */
export function getTrackedConnection(
  conn: object,
): { connId: number; pool: string; checkedOutAt: number; stack: string } | undefined {
  return trackedConnections.get(conn)
}
